File: backend/chesscom_profile.py
from urllib.parse import quote

import requests
from fastapi import APIRouter, HTTPException, Query

router = APIRouter()

CHESS_COM_HEADERS = {
    "User-Agent": "ChessGameReview/1.0 (https://github.com/mrprohack/chess_rev)",
    "Accept": "application/json",
}


def _request_json(session: requests.Session, url: str, *, not_found_detail: str | None = None):
    response = session.get(url, headers=CHESS_COM_HEADERS, timeout=10)
    if response.status_code == 404:
        raise HTTPException(status_code=404, detail=not_found_detail or "Chess.com data not found")
    if response.status_code == 429:
        raise HTTPException(status_code=429, detail="Chess.com rate limit reached. Try again shortly.")
    response.raise_for_status()
    return response.json()


def _rating(stats: dict, key: str):
    value = stats.get(key, {}).get("last", {}).get("rating")
    return value if isinstance(value, (int, float)) else None


def _normalize_game(game: dict):
    return {
        "uuid": game.get("uuid"),
        "url": game.get("url"),
        "end_time": game.get("end_time"),
        "time_class": game.get("time_class"),
        "rated": bool(game.get("rated")),
        "white": {
            "username": game.get("white", {}).get("username"),
            "rating": game.get("white", {}).get("rating"),
            "result": game.get("white", {}).get("result"),
        },
        "black": {
            "username": game.get("black", {}).get("username"),
            "rating": game.get("black", {}).get("rating"),
            "result": game.get("black", {}).get("result"),
        },
    }


@router.get("/api/chesscom/profile/{username}")
def get_chess_com_profile(username: str, limit: int = Query(default=12, ge=1, le=20)):
    cleaned = username.strip()
    if not cleaned or len(cleaned) > 50:
        raise HTTPException(status_code=400, detail="Invalid Chess.com username")

    encoded = quote(cleaned, safe="")
    base = f"https://api.chess.com/pub/player/{encoded}"

    try:
        with requests.Session() as session:
            profile = _request_json(
                session,
                base,
                not_found_detail="Chess.com profile not found",
            )
            stats = _request_json(
                session,
                f"{base}/stats",
                not_found_detail="Chess.com profile stats not found",
            )
            archives = _request_json(
                session,
                f"{base}/games/archives",
                not_found_detail="Chess.com game archive not found",
            )

            games = []
            archive_urls = archives.get("archives", [])
            for archive_url in reversed(archive_urls[-3:]):
                archive = _request_json(session, archive_url)
                month_games = [
                    _normalize_game(game)
                    for game in archive.get("games", [])
                    if game.get("rules", "chess") == "chess" and game.get("url")
                ]
                month_games.sort(key=lambda item: item.get("end_time") or 0, reverse=True)
                games.extend(month_games)
                if len(games) >= limit:
                    break

            games = games[:limit]

        return {
            "username": profile.get("username") or cleaned,
            "name": profile.get("name"),
            "avatar": profile.get("avatar"),
            "url": profile.get("url") or f"https://www.chess.com/member/{encoded}",
            "status": profile.get("status"),
            "ratings": {
                "rapid": _rating(stats, "chess_rapid"),
                "blitz": _rating(stats, "chess_blitz"),
                "bullet": _rating(stats, "chess_bullet"),
            },
            "games": games,
        }
    except HTTPException:
        raise
    except requests.Timeout:
        raise HTTPException(status_code=504, detail="Chess.com timed out") from None
    except requests.RequestException:
        raise HTTPException(status_code=502, detail="Could not reach Chess.com") from None

File: mobile/lib/data/models/chesscom_profile.dart
class ChessComProfile {
  const ChessComProfile({
    required this.username,
    required this.ratings,
    required this.games,
    this.name,
    this.avatar,
    this.url,
    this.status,
  });

  final String username;
  final String? name;
  final String? avatar;
  final String? url;
  final String? status;
  final ChessComRatings ratings;
  final List<RecentGame> games;

  factory ChessComProfile.fromJson(Map<String, dynamic> json) {
    final rawRatings = json['ratings'];
    final rawGames = json['games'];
    if (rawRatings is! Map || rawGames is! List) {
      throw const FormatException('profile ratings and games are required');
    }
    return ChessComProfile(
      username: _requiredString(json['username'], 'username'),
      name: _optionalString(json['name']),
      avatar: _optionalString(json['avatar']),
      url: _optionalString(json['url']),
      status: _optionalString(json['status']),
      ratings: ChessComRatings.fromJson(_asMap(rawRatings)),
      games: rawGames
          .map((item) => RecentGame.fromJson(_asMap(item)))
          .toList(growable: false),
    );
  }

  Map<String, dynamic> toJson() => {
    'username': username,
    'name': name,
    'avatar': avatar,
    'url': url,
    'status': status,
    'ratings': ratings.toJson(),
    'games': games.map((game) => game.toJson()).toList(growable: false),
  };
}

class ChessComRatings {
  const ChessComRatings({this.rapid, this.blitz, this.bullet});

  final int? rapid;
  final int? blitz;
  final int? bullet;

  factory ChessComRatings.fromJson(Map<String, dynamic> json) {
    return ChessComRatings(
      rapid: _toInt(json['rapid']),
      blitz: _toInt(json['blitz']),
      bullet: _toInt(json['bullet']),
    );
  }

  Map<String, dynamic> toJson() => {
    'rapid': rapid,
    'blitz': blitz,
    'bullet': bullet,
  };
}

class RecentGame {
  const RecentGame({
    required this.url,
    required this.white,
    required this.black,
    this.uuid,
    this.endTime,
    this.timeClass,
    this.rated = false,
  });

  final String? uuid;
  final String url;
  final int? endTime;
  final String? timeClass;
  final bool rated;
  final RecentPlayer white;
  final RecentPlayer black;

  factory RecentGame.fromJson(Map<String, dynamic> json) {
    return RecentGame(
      uuid: _optionalString(json['uuid']),
      url: _requiredString(json['url'], 'url'),
      endTime: _toInt(json['end_time']),
      timeClass: _optionalString(json['time_class']),
      rated: json['rated'] == true,
      white: RecentPlayer.fromJson(_asMap(json['white'])),
      black: RecentPlayer.fromJson(_asMap(json['black'])),
    );
  }

  Map<String, dynamic> toJson() => {
    'uuid': uuid,
    'url': url,
    'end_time': endTime,
    'time_class': timeClass,
    'rated': rated,
    'white': white.toJson(),
    'black': black.toJson(),
  };
}

class RecentPlayer {
  const RecentPlayer({required this.username, this.rating, this.result});

  final String username;
  final int? rating;
  final String? result;

  factory RecentPlayer.fromJson(Map<String, dynamic> json) {
    return RecentPlayer(
      username: _requiredString(json['username'], 'username'),
      rating: _toInt(json['rating']),
      result: _optionalString(json['result']),
    );
  }

  Map<String, dynamic> toJson() => {
    'username': username,
    'rating': rating,
    'result': result,
  };
}

Map<String, dynamic> _asMap(Object? value) {
  if (value is Map<String, dynamic>) {
    return value;
  }
  if (value is Map) {
    return value.map((key, item) => MapEntry(key.toString(), item));
  }
  throw const FormatException('expected object');
}

String _requiredString(Object? value, String label) {
  if (value is String && value.isNotEmpty) {
    return value;
  }
  throw FormatException('$label is required');
}

String? _optionalString(Object? value) {
  if (value == null) {
    return null;
  }
  final string = value.toString();
  return string.isEmpty ? null : string;
}

int? _toInt(Object? value) {
  if (value is int) {
    return value;
  }
  if (value is num) {
    return value.toInt();
  }
  return int.tryParse(value?.toString() ?? '');
}
